internal/pkg/util/shell/interpreter/interpreter.go

```diff
@@ -0,0 +1,301 @@
+func (s *Shell) internalOpenHandler() interp.OpenHandlerFunc {
+	return func(ctx context.Context, path string, flag int, perm os.FileMode) (io.ReadWriteCloser, error) {
+		mc := interp.HandlerCtx(ctx)
+		if !filepath.IsAbs(path) {
+			path = filepath.Join(mc.Dir, path)
+		}
+		if handler, ok := s.openHandlers[path]; ok {
+			return handler(path, flag, perm)
+		}
+		return os.OpenFile(path, flag, perm)
+	}
+}
+
+// RegisterShellBuiltin registers a shell interpreter builtin.
+func (s *Shell) RegisterShellBuiltin(name string, builtin ShellBuiltin) {
+	if s.shellBuiltins == nil {
+		s.shellBuiltins = make(map[string]ShellBuiltin)
+	}
+	s.shellBuiltins[name] = builtin
+}
+
+// RegisterOpenHandler registers a shell interpreter file handler.
+func (s *Shell) RegisterOpenHandler(path string, handler OpenHandler) {
+	if s.openHandlers == nil {
+		s.openHandlers = make(map[string]OpenHandler)
+	}
+	s.openHandlers[path] = handler
+}
+
+// LookPath returns the absolute path for the command passed in argument
+// within the context of the shell interpreter.
+func (s *Shell) LookPath(ctx context.Context, cmd string) (string, error) {
+	hc := interp.HandlerCtx(ctx)
+	return interp.LookPath(hc.Env, cmd)
+}
+
+// Run runs the shell interpreter.
+func (s *Shell) Run() error {
+	ctx := context.TODO()
+
+	parser := syntax.NewParser()
+	node, err := parser.Parse(s.reader, s.name)
+	if err != nil {
+		return fmt.Errorf("while parsing script: %s", err)
+	}
+
+	if err := s.runner.Run(ctx, node); err != nil {
+		if status, ok := interp.IsExitStatus(err); ok {
+			s.status = status
+		}
+		return err
+	}
+
+	return nil
+}
+
+// Status returns the exit code status of the shell interpreter.
+func (s *Shell) Status() uint8 {
+	return s.status
+}
+
+// nonExportedEnv allows to initialize shell interpreter with all environment
+// variables set as non exported variables.
+type nonExportedEnv struct {
+	envs map[string]expand.Variable
+}
+
+// newNonExportedEnv returns a localEnv instance associated to environment
+// passed in argument.
+func newNonExportedEnv(env []string) nonExportedEnv {
+	local := nonExportedEnv{
+		envs: make(map[string]expand.Variable),
+	}
+	for _, e := range env {
+		e := strings.SplitN(e, "=", 2)
+		local.envs[e[0]] = expand.Variable{Str: e[1], Kind: expand.String}
+	}
+	return local
+}
+
+// Get returns the named shell environment variable.
+func (e nonExportedEnv) Get(name string) expand.Variable {
+	if vr, ok := e.envs[name]; ok {
+		return vr
+	}
+	return expand.Variable{}
+}
+
+// Each iterates over all environment variables by calling the
+// function passed in argument for each variables.
+func (e nonExportedEnv) Each(fn func(name string, vr expand.Variable) bool) {
+	for name, vr := range e.envs {
+		if !fn(name, vr) {
+			return
+		}
+	}
+}
+
+// EvaluateEnv evaluates the environment variable script and returns
+// the list of variables set in the script. Command execution is disabled
```

Comment:
So this will break any container definition files where people rely on old behavior where `%environment` *can* execute things inside the container... 

```
%environment
	export STARTED=$(date)
```

.. not opposed to that really, but do we know how many people use it?

Reply:
`EvaluateEnv` is used only for `--env-file`, I choose to disable command execution, sourcing and redirection to reduce potential abuse, but everything in `%environment` section is running as usual within the shell interpreter without any restriction